File: python/tests/test_order_book_window.py
"""Tests for ``flox_py.OrderBook`` tick handling and delta-fed window tracking.

Mirrors tests/test_nlevel_order_book_window.cpp through the Python binding,
which reads the same book the engine does.

Run from repo root:
    PYTHONPATH=build/python python3 python/tests/test_order_book_window.py
"""
from __future__ import annotations

import sys
import unittest
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
for cand in ("build/python", "build-py312/python"):
    p = REPO_ROOT / cand
    if p.is_dir():
        sys.path.insert(0, str(p))
        break

import numpy as np  # noqa: E402

import flox_py  # noqa: E402


def _arr(values) -> np.ndarray:
    return np.array(values, dtype=np.float64)


class TickSizeValidationTests(unittest.TestCase):
    def test_zero_tick_size_is_rejected(self) -> None:
        with self.assertRaises(ValueError):
            flox_py.OrderBook(0.0)

    def test_negative_tick_size_is_rejected(self) -> None:
        with self.assertRaises(ValueError):
            flox_py.OrderBook(-0.01)

    def test_positive_tick_size_is_accepted(self) -> None:
        book = flox_py.OrderBook(0.01)
        self.assertIsNone(book.best_bid())


class TinyTickSizeTests(unittest.TestCase):
    """A 1e-8 tick is a raw divisor of 1 at a price scale of 1e8, which is what
    the sub-cent pairs quote in. Every level used to collapse onto one index."""

    def test_levels_stay_apart_at_one_satoshi(self) -> None:
        book = flox_py.OrderBook(1e-8)
        book.apply_snapshot(_arr([1e-5]), _arr([1.0]),
                            _arr([1.002e-5]), _arr([1.0]))

        self.assertAlmostEqual(book.best_bid(), 1e-5, places=12)
        self.assertAlmostEqual(book.best_ask(), 1.002e-5, places=12)
        self.assertFalse(book.is_crossed())
        self.assertAlmostEqual(book.spread(), 2e-8, places=12)
        self.assertAlmostEqual(book.mid(), 1.001e-5, places=12)

    def test_levels_stay_apart_at_two_satoshi(self) -> None:
        book = flox_py.OrderBook(2e-8)
        book.apply_snapshot(_arr([1e-5]), _arr([1.0]),
                            _arr([1.002e-5]), _arr([1.0]))

        self.assertAlmostEqual(book.best_bid(), 1e-5, places=12)
        self.assertAlmostEqual(book.best_ask(), 1.002e-5, places=12)
        self.assertFalse(book.is_crossed())


class MidTests(unittest.TestCase):
    def test_mid_is_the_average_of_both_sides(self) -> None:
        book = flox_py.OrderBook(5e-8)
        book.apply_snapshot(_arr([1e-5]), _arr([1.0]),
                            _arr([1.01e-5]), _arr([1.0]))

        expected = (book.best_bid() + book.best_ask()) / 2
        self.assertAlmostEqual(book.mid(), expected, places=12)


class DeltaDriftTests(unittest.TestCase):
    """A delta feed re-sends a snapshot only on connect and on a sequence gap,
    so the tick window has to follow the market between the two."""

    def test_book_survives_a_long_delta_only_drift(self) -> None:
        tick = 0.1
        book = flox_py.OrderBook(tick)

        bid_top, ask_top = 59999.9, 60000.1
        depth = 10
        book.apply_snapshot(
            _arr([round(bid_top - k * tick, 8) for k in range(depth)]),
            _arr([1.0] * depth),
            _arr([round(ask_top + k * tick, 8) for k in range(depth)]),
            _arr([1.0] * depth))

        for step in range(1, 601):
            new_bid = round(bid_top + step * tick, 8)
            new_ask = round(ask_top + step * tick, 8)
            book.apply_delta(
                _arr([new_bid, round(new_bid - depth * tick, 8)]), _arr([1.0, 0.0]),
                _arr([round(new_ask + (depth - 1) * tick, 8), round(new_ask - tick, 8)]),
                _arr([1.0, 0.0]))

            self.assertIsNotNone(book.best_bid(), f"bid side went dark at step {step}")
            self.assertIsNotNone(book.best_ask(), f"ask side went dark at step {step}")
            self.assertAlmostEqual(book.best_bid(), new_bid, places=6,
                                   msg=f"at step {step}")
            self.assertAlmostEqual(book.best_ask(), new_ask, places=6,
                                   msg=f"at step {step}")


if __name__ == "__main__":
    unittest.main()

File: node/test/test_order_book_window.js
// OrderBook tick handling and delta-fed window tracking through the Node
// binding, which reads the same C++ book the engine does.

const assert = require('assert');
const flox = require('../index.js');

let passed = 0;
function check(cond, msg) {
  assert.ok(cond, msg);
  passed++;
}
function throws(fn, msg) {
  let threw = null;
  try { fn(); } catch (e) { threw = e; }
  assert.ok(threw, msg + ' (expected a throw, got none)');
  passed++;
}
function near(a, b, eps, msg) {
  assert.ok(Math.abs(a - b) <= eps, `${msg} (got ${a}, want ${b})`);
  passed++;
}

// ── Tick size validation ──────────────────────────────────────────────
// A zero tick divided by zero while building the reciprocal and left a book
// that was dead for the rest of its life without a single diagnostic.
throws(() => new flox.OrderBook(0), 'zero tick size is rejected');
throws(() => new flox.OrderBook(-0.01), 'negative tick size is rejected');
check(new flox.OrderBook(0.01) !== null, 'a positive tick size is accepted');

// ── Sub-cent tick sizes ───────────────────────────────────────────────
// A 1e-8 tick is a raw divisor of 1 at a price scale of 1e8. Every level used
// to collapse onto one index, which left the book permanently crossed.
{
  const book = new flox.OrderBook(1e-8);
  book.applySnapshot(Float64Array.from([1e-5]), Float64Array.from([1.0]),
                     Float64Array.from([1.002e-5]), Float64Array.from([1.0]));
  near(book.bestBid(), 1e-5, 1e-14, 'best bid at a 1e-8 tick');
  near(book.bestAsk(), 1.002e-5, 1e-14, 'best ask at a 1e-8 tick');
  check(book.isCrossed() === false, 'a normal book is not crossed at a 1e-8 tick');
  near(book.mid(), 1.001e-5, 1e-14, 'mid at a 1e-8 tick');
  book.clear();
}

// ── Delta-only drift ──────────────────────────────────────────────────
// A delta feed re-sends a snapshot only on connect and on a sequence gap, so
// the tick window has to follow the market between the two. It did not: about
// $25 of movement emptied one side of the book and $30 emptied both.
{
  const tick = 0.1;
  const depth = 10;
  const book = new flox.OrderBook(tick);
  const bidTop = 59999.9;
  const askTop = 60000.1;

  const round8 = (v) => Math.round(v * 1e8) / 1e8;
  const bids = [], asks = [], qty = [];
  for (let k = 0; k < depth; ++k) {
    bids.push(round8(bidTop - k * tick));
    asks.push(round8(askTop + k * tick));
    qty.push(1.0);
  }
  book.applySnapshot(Float64Array.from(bids), Float64Array.from(qty),
                     Float64Array.from(asks), Float64Array.from(qty));

  let lastBid = 0, lastAsk = 0;
  for (let step = 1; step <= 600; ++step) {
    const newBid = round8(bidTop + step * tick);
    const newAsk = round8(askTop + step * tick);
    book.applyDelta(
      Float64Array.from([newBid, round8(newBid - depth * tick)]),
      Float64Array.from([1.0, 0.0]),
      Float64Array.from([round8(newAsk + (depth - 1) * tick), round8(newAsk - tick)]),
      Float64Array.from([1.0, 0.0]));

    assert.ok(book.bestBid() !== null, `bid side went dark at step ${step}`);
    assert.ok(book.bestAsk() !== null, `ask side went dark at step ${step}`);
    lastBid = book.bestBid();
    lastAsk = book.bestAsk();
    assert.ok(Math.abs(lastBid - newBid) < 1e-6, `best bid wrong at step ${step}`);
    assert.ok(Math.abs(lastAsk - newAsk) < 1e-6, `best ask wrong at step ${step}`);
  }
  passed++;
  near(lastBid, round8(bidTop + 600 * tick), 1e-6, 'best bid after $60 of drift');
  near(lastAsk, round8(askTop + 600 * tick), 1e-6, 'best ask after $60 of drift');
  book.clear();
}

console.log(`test_order_book_window: ${passed} checks passed`);
